README
NAME
    SHARYANTO::Module::Path - Get path to locally installed Perl module

VERSION
    This document describes version 0.21 of SHARYANTO::Module::Path (from
    Perl distribution SHARYANTO-Module-Path), released on 2014-12-02.

SYNOPSIS
     use SHARYANTO::Module::Path 'module_path', 'pod_path';

     $path = module_path(module=>'Test::More');
     if (defined($path)) {
       print "Test::More found at $path\n";
     } else {
       print "Danger Will Robinson!\n";
     }

     # find all found modules, as well as .pmc and .pod files
     @path = module_path(module=>'Foo::Bar', all=>1, find_pmc=>1, find_pod=>1);

     # just a shortcut for module_path(module=>'Foo',
                                       find_pm=>0, find_pmc=>0, find_pod=>1);
     $path = pod_path(module=>'Foo');

DESCRIPTION
    This module is a fork of Module::Path. It contains features that are not
    (or have not been accepted) in the original module, namely: finding all
    matches instead of the first found match, and finding .pmc/.pod in
    addition to .pm files. There is also a difference of behavior: no
    abs_path() or symlink resolving is being done by default because I think
    that's the sensible default (doing abs_path() or resolving symlinks will
    sometimes fail or expose filesystem quirks that we might not want to
    deal with at all). However, an "abs" bool option is provided if a user
    wants to do that.

    This module has also diverged by introducing a different interface since
    v0.14.

    References:

    *   <https://github.com/neilbowers/Module-Path/issues/6>

    *   <https://github.com/neilbowers/Module-Path/issues/7>

    *   <https://github.com/neilbowers/Module-Path/issues/10>

FUNCTIONS
  module_path(%args) -> array|str
    Get path to locally installed Perl module.

    Search @INC (reference entries are skipped) and return path(s) to Perl
    module files with the requested name.

    This function is like the one from "Module::Path", except with a
    different interface and more options (finding all matches instead of the
    first, the option of not absolutizing paths, finding ".pmc" & ".pod"
    files, finding module prefixes).

    Arguments ('*' denotes required arguments):

    *   abs => *bool* (default: 0)

        Whether to return absolute paths.

    *   all => *bool* (default: 0)

        Return all results instead of just the first.

    *   find_pm => *bool* (default: 1)

        Whether to find .pm files.

    *   find_pmc => *bool* (default: 1)

        Whether to find .pmc files.

    *   find_pod => *bool* (default: 0)

        Whether to find .pod files.

    *   find_prefix => *bool* (default: 1)

        Whether to find module prefixes.

    *   module* => *str*

        Module name to search.

    Return value:

     (any)

  pod_path(%args) -> array|str
    Find path to Perl POD files.

    Shortcut for "module_path(..., find_pm=>0, find_pmc=>0, find_pod=>1,
    find_prefix=>1, )".

    Arguments ('*' denotes required arguments):

    *   abs => *bool* (default: 0)

        Whether to return absolute paths.

    *   all => *bool* (default: 0)

        Return all results instead of just the first.

    *   module* => *str*

        Module name to search.

    Return value:

     (any)

SEE ALSO
    SHARYANTO

    Module::Path

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/SHARYANTO-Module-Path>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-Alt-Module-Path-SHARYANTO>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=SHARYANTO-Module-Path
    >

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for SHARYANTO-Module-Path

0.21    2014-12-02 (PERLANCAR)

	- No functional changes.

	- Fix/sync Rinci metadata (find_prefix is by default 0).


0.20     2014-08-26 (SHARYANTO)

         - [Removed] Remove completion-related routines (moved to App-PMUtils).

         - Add note in Description about why this module exists.


0.19     2014-06-25 (SHARYANTO)

         - Use case-insensitive module completion, for convenience.


0.18     2014-06-25 (SHARYANTO)

         - Fix typo.

         - Use Complete::Util (renamed from SHARYANTO::Complete::Util).


0.17     2014-06-24 (SHARYANTO)

         - No functional changes.

         - Fix test to avoid failure [CT].


0.16     2014-06-24 (SHARYANTO)

         - A couple of bug fixes and tweaks.


0.15     2014-06-23 (SHARYANTO)

         - Add completion routine 'module' argument.


0.14     2014-06-22 (SHARYANTO)

         [INCOMPATIBLE CHANGES]

         - Rename dist/module from Alt::Module::Path::SHARYANTO to
           SHARYANTO::Module::Path because I want to depend on this module on
           CPAN (and don't want to force others to have to replace the original
           Module::Path by installing this dist).

         - Change interface (using hash argument) so it becomes standard Rinci
           stuff.

         [REMOVED FEATURES]

         - Remove command-line script 'mpath' to avoid clash with the one from
           Module-Path. Will replace with App-PMUtils' pmpath.


0.13.2   2014-06-22 (SHARYANTO)

         - [Bugfix] Fix logic bug which made --find-prefix negates other
           --find-*.

         - Add short options: -a (for --all), -P (for --abs).


0.13.1   2014-06-22 (SHARYANTO)

         - Add option: find_prefix (--find-prefix, -p).


0.13     2014-04-25 (SHARYANTO)

         - No functional changes.

         - Increase version number to match original's (0.13).


0.02     2014-04-12 (SHARYANTO)

         - Some tweaks to bin/mpath: Make --pod equal to --find-pod --nofind-pm
           --nofind-pmc. Turn on full by default if there are more than one
           argument. Use tab as separator character.


0.01     2014-04-09 (SHARYANTO)

         - First release.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Get path to locally installed Perl module",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "SHARYANTO-Module-Path",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "SHARYANTO::Module::Path",
  "PREREQ_PM" => {
    "Perinci::Sub::Util" => "0.38",
    "Rinci" => "v1.1.0"
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.21",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Perinci::Sub::Util" => "0.38",
  "Rinci" => "v1.1.0",
  "Test::More" => "0.98"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/SHARYANTO/Module/Path.pm
package SHARYANTO::Module::Path;

use 5.010001;
use strict;
use warnings;

use Perinci::Sub::Util qw(gen_modified_sub);

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(module_path pod_path);

our $VERSION = '0.21'; # VERSION
our $DATE = '2014-12-02'; # DATE

my $SEPARATOR;

our %SPEC;

BEGIN {
    if ($^O =~ /^(dos|os2)/i) {
        $SEPARATOR = '\\';
    } elsif ($^O =~ /^MacOS/i) {
        $SEPARATOR = ':';
    } else {
        $SEPARATOR = '/';
    }
}

$SPEC{module_path} = {
    v => 1.1,
    summary => 'Get path to locally installed Perl module',
    description => <<'_',

Search `@INC` (reference entries are skipped) and return path(s) to Perl module
files with the requested name.

This function is like the one from `Module::Path`, except with a different
interface and more options (finding all matches instead of the first, the option
of not absolutizing paths, finding `.pmc` & `.pod` files, finding module
prefixes).

_
    args => {
        module => {
            summary => 'Module name to search',
            schema  => 'str*',
            req     => 1,
            pos     => 0,
        },
        find_pm => {
            summary => 'Whether to find .pm files',
            schema  => 'bool',
            default => 1,
        },
        find_pmc => {
            summary => 'Whether to find .pmc files',
            schema  => 'bool',
            default => 1,
        },
        find_pod => {
            summary => 'Whether to find .pod files',
            schema  => 'bool',
            default => 0,
        },
        find_prefix => {
            summary => 'Whether to find module prefixes',
            schema  => 'bool',
            default => 0,
        },
        all => {
            summary => 'Return all results instead of just the first',
            schema  => 'bool',
            default => 0,
        },
        abs => {
            summary => 'Whether to return absolute paths',
            schema  => 'bool',
            default => 0,
        },
    },
    result => {
        schema => ['any' => of => ['str*', ['array*' => of => 'str*']]],
    },
    result_naked => 1,
};
sub module_path {
    my %args = @_;

    my $module = $args{module} or die "Please specify module";

    $args{abs}         //= 0;
    $args{all}         //= 0;
    $args{find_pm}     //= 1;
    $args{find_pmc}    //= 1;
    $args{find_pod}    //= 0;
    $args{find_prefix} //= 0;

    require Cwd if $args{abs};

    my @res;
    my $add = sub { push @res, $args{abs} ? Cwd::abs_path($_[0]) : $_[0] };

    my $relpath;

    ($relpath = $module) =~ s/::/$SEPARATOR/g;
    $relpath =~ s/\.(pm|pmc|pod)\z//i;

    foreach my $dir (@INC) {
        next if not defined($dir);
        next if ref($dir);

        my $prefix = $dir . $SEPARATOR . $relpath;
        if ($args{find_pmc}) {
            my $file = $prefix . ".pmc";
            if (-f $file) {
                $add->($file);
                last unless $args{all};
            }
        }
        if ($args{find_pm}) {
            my $file = $prefix . ".pm";
            if (-f $file) {
                $add->($file);
                last unless $args{all};
            }
        }
        if ($args{find_pod}) {
            my $file = $prefix . ".pod";
            if (-f $file) {
                $add->($file);
                last unless $args{all};
            }
        }
        if ($args{find_prefix}) {
            if (-d $prefix) {
                $add->($prefix);
                last unless $args{all};
            }
        }
    }

    if ($args{all}) {
        return \@res;
    } else {
        return @res ? $res[0] : undef;
    }
}

gen_modified_sub(
    output_name => 'pod_path',
    base_name   => 'module_path',
    summary     => 'Find path to Perl POD files',
    description => <<'_',

Shortcut for `module_path(..., find_pm=>0, find_pmc=>0, find_pod=>1,
find_prefix=>1, )`.

_
    remove_args => [qw/find_pm find_pmc find_pod find_prefix/],
    output_code => sub {
        my %args = @_;
        module_path(
            %args, find_pm=>0, find_pmc=>0, find_pod=>1, find_prefix=>0);
    },
);

1;
# ABSTRACT: Get path to locally installed Perl module

__END__

=pod

=encoding UTF-8

=head1 NAME

SHARYANTO::Module::Path - Get path to locally installed Perl module

=head1 VERSION

This document describes version 0.21 of SHARYANTO::Module::Path (from Perl distribution SHARYANTO-Module-Path), released on 2014-12-02.

=head1 SYNOPSIS

 use SHARYANTO::Module::Path 'module_path', 'pod_path';

 $path = module_path(module=>'Test::More');
 if (defined($path)) {
   print "Test::More found at $path\n";
 } else {
   print "Danger Will Robinson!\n";
 }

 # find all found modules, as well as .pmc and .pod files
 @path = module_path(module=>'Foo::Bar', all=>1, find_pmc=>1, find_pod=>1);

 # just a shortcut for module_path(module=>'Foo',
                                   find_pm=>0, find_pmc=>0, find_pod=>1);
 $path = pod_path(module=>'Foo');

=head1 DESCRIPTION

This module is a fork of L<Module::Path>. It contains features that are not (or
have not been accepted) in the original module, namely: finding all matches
instead of the first found match, and finding .pmc/.pod in addition to .pm
files. There is also a difference of behavior: no abs_path() or symlink
resolving is being done by default because I think that's the sensible default
(doing abs_path() or resolving symlinks will sometimes fail or expose filesystem
quirks that we might not want to deal with at all). However, an C<abs> bool
option is provided if a user wants to do that.

This module has also diverged by introducing a different interface since v0.14.

References:

=over

=item * L<https://github.com/neilbowers/Module-Path/issues/6>

=item * L<https://github.com/neilbowers/Module-Path/issues/7>

=item * L<https://github.com/neilbowers/Module-Path/issues/10>

=back

=head1 FUNCTIONS


=head2 module_path(%args) -> array|str

Get path to locally installed Perl module.

Search C<@INC> (reference entries are skipped) and return path(s) to Perl module
files with the requested name.

This function is like the one from C<Module::Path>, except with a different
interface and more options (finding all matches instead of the first, the option
of not absolutizing paths, finding C<.pmc> & C<.pod> files, finding module
prefixes).

Arguments ('*' denotes required arguments):

=over 4

=item * B<abs> => I<bool> (default: 0)

Whether to return absolute paths.

=item * B<all> => I<bool> (default: 0)

Return all results instead of just the first.

=item * B<find_pm> => I<bool> (default: 1)

Whether to find .pm files.

=item * B<find_pmc> => I<bool> (default: 1)

Whether to find .pmc files.

=item * B<find_pod> => I<bool> (default: 0)

Whether to find .pod files.

=item * B<find_prefix> => I<bool> (default: 1)

Whether to find module prefixes.

=item * B<module>* => I<str>

Module name to search.

=back

Return value:

 (any)


=head2 pod_path(%args) -> array|str

Find path to Perl POD files.

Shortcut for C<< module_path(..., find_pm=E<gt>0, find_pmc=E<gt>0, find_pod=E<gt>1,
find_prefix=E<gt>1, ) >>.

Arguments ('*' denotes required arguments):

=over 4

=item * B<abs> => I<bool> (default: 0)

Whether to return absolute paths.

=item * B<all> => I<bool> (default: 0)

Return all results instead of just the first.

=item * B<module>* => I<str>

Module name to search.

=back

Return value:

 (any)

=head1 SEE ALSO

L<SHARYANTO>

L<Module::Path>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/SHARYANTO-Module-Path>.

=head1 SOURCE

Source repository is at L<https://github.com/perlancar/perl-Alt-Module-Path-SHARYANTO>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=SHARYANTO-Module-Path>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

perlancar <perlancar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by perlancar@cpan.org.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
